### PR TITLE
Add async/await helper for CollectionReference.add()/Firestore.loadBundle()

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This adds an async method for `CollectionReference.addDocument(data:)`. This method does not automatically work with async/await since it returns a DocumentReference.

### CODE REVIEW

**Comment by mortenbekditlevsen** [comment]
Just a thought:

The only reason that this API isn't automatically synthesised from the Objective-C code is that the method contains both a return value and a completion handler:

```
- (FIRDocumentReference *)addDocumentWithData:(NSDictionary<NSString *, id> *)data
                                   completion:
                                       (nullable void (^)(NSError *_Nullable error))completion {
  ParsedSetData parsed = [self.firestore.dataReader parsedSetData:data];
  DocumentReference docRef =
      self.reference.AddDocument(std::move(parsed), MakeCallback(completion));
  return [[FIRDocumentReference alloc] initWithReference:std::move(docRef)];
}
```

If the function signature looked like this: 
```
- (void)addDocumentWithData:(NSDictionary<NSString *, id> *)data
                                   completion:
                                       (nullable void (^)(FIRDocumentReference *_Nullable reference, NSError *_Nullable error))completion 
```
then I assume it would get synthesised.
Might it be worth adding that overload directly to Objective-C?

**Comment by paulb777** [comment]
@mortenbekditlevsen I had the same thought when I was working on https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseStorageSwift/Sources/AsyncAwait.swift, but as far as I can tell, Objective C does not do overloading when only the return type is changed. I believe only the parameters are part of the function signature.

**Comment by mortenbekditlevsen** [comment]
Thanks, Paul! That's good to know! 😊

**Comment by schmidt-sebastian** [comment]
@mortenbekditlevsen I tried you suggestion as both the return type and the type of the closure changed, but I am still getting a duplicate function definition error. So ++ on what @paulb777 said.

**Comment by mortenbekditlevsen** [comment]
Thanks for the update, and apologies for the wild goose chase.

File: Firestore/Swift/Source/AsyncAwait/CollectionReference+AsyncAwait.swift [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ * limitations under the License.
+ */
+
+import FirebaseFirestore
+import Foundation
+
+#if compiler(>=5.5) && canImport(_Concurrency)
+  @available(iOS 15, tvOS 15, macOS 12, watchOS 8, *)
+  public extension CollectionReference {
+    /**
```
Comment (paulb777) [inline Firestore/Swift/Source/AsyncAwait/CollectionReference+AsyncAwait.swift:23]: Use triple slash for doc comments - like https://github.com/firebase/firebase-ios-sdk/blob/master/FirebaseFunctionsSwift/Sources/Codable/Callable%2BCodable.swift#L20

File: Firestore/Swift/Tests/API/BasicCompileTests.swift [hunk 2]
```diff
@@ -472,3 +472,24 @@ func terminateDb(database db: Firestore) {
...
+    try await db.waitForPendingWrites()
+    try await db.terminate()
+    try await db.runTransaction { _, _ in
+      0
+    }
+
+    let collection = db.collection("coll")
+    let document = try await collection.addDocument(data: [:])
+
+    try await document.setData([:])
```
Comment (paulb777) [inline Firestore/Swift/Tests/API/BasicCompileTests.swift:490]: Add the mergeFields and merge variations of the setData API.

File: Firestore/Swift/Tests/API/BasicCompileTests.swift [hunk 3]
```diff
@@ -472,3 +472,24 @@ func terminateDb(database db: Firestore) {
...
+    try await document.updateData([:])
+    try await document.delete()
+    try await document.getDocument()
```
Comment (paulb777) [inline Firestore/Swift/Tests/API/BasicCompileTests.swift:493]: Add the source parameter variation of getDocument

File: Firestore/Swift/Tests/API/BasicCompileTests.swift [hunk 4]
```diff
@@ -472,3 +472,24 @@ func terminateDb(database db: Firestore) {
     }
   }
 }
+
+#if swift(>=5.5)
+  @available(iOS 15, tvOS 15, macOS 12, watchOS 8, *)
+  func testAsyncAwait(database db: Firestore) async throws {
+    try await db.enableNetwork()
+    try await db.disableNetwork()
+    try await db.waitForPendingWrites()
```
Comment (paulb777) [inline Firestore/Swift/Tests/API/BasicCompileTests.swift:481]: Missing `clearPersistence`

File: Firestore/Swift/Tests/API/BasicCompileTests.swift [hunk 5]
```diff
@@ -472,3 +472,24 @@ func terminateDb(database db: Firestore) {
...
+    }
+
+    let collection = db.collection("coll")
+    let document = try await collection.addDocument(data: [:])
+
+    try await document.setData([:])
+    try await document.updateData([:])
+    try await document.delete()
+    try await document.getDocument()
+  }
```
Comment (paulb777) [inline Firestore/Swift/Tests/API/BasicCompileTests.swift:494]: Missing `getQuery`

Should there be an `async` variation for the `loadBundle` APIs?

File: Firestore/Swift/Tests/API/BasicCompileTests.swift (hunk 3 above)
Comment (paulb777) [inline Firestore/Swift/Tests/API/BasicCompileTests.swift:493]: Add query.getDocuments APIs

**Review by paulb777 — COMMENTED** [review]
If you weren't intending to get full coverage in this PR, feel free to defer those comments.

If so, you can also update https://github.com/firebase/firebase-ios-sdk/blob/master/SwiftDashboard.md

I'm not sure if we need an API review, when we're filling in gaps like this? @ryanwilson

**Review by schmidt-sebastian — COMMENTED** [review]
I have addressed all feedback and also added a loadBundle overload.

File: Firestore/Swift/Source/AsyncAwait/CollectionReference+AsyncAwait.swift (hunk 1 above)
Comment (schmidt-sebastian) [reply Firestore/Swift/Source/AsyncAwait/CollectionReference+AsyncAwait.swift]: Done

File: Firestore/Swift/Tests/API/BasicCompileTests.swift (hunk 4 above)
Comment (schmidt-sebastian) [reply Firestore/Swift/Tests/API/BasicCompileTests.swift:481]: Done

File: Firestore/Swift/Tests/API/BasicCompileTests.swift (hunk 2 above)
Comment (schmidt-sebastian) [reply Firestore/Swift/Tests/API/BasicCompileTests.swift:500]: Done

File: Firestore/Swift/Tests/API/BasicCompileTests.swift (hunk 3 above)
Comment (schmidt-sebastian) [reply Firestore/Swift/Tests/API/BasicCompileTests.swift:505]: Done

Comment (schmidt-sebastian) [reply Firestore/Swift/Tests/API/BasicCompileTests.swift:505]: Done

File: Firestore/Swift/Tests/API/BasicCompileTests.swift (hunk 5 above)
Comment (schmidt-sebastian) [reply Firestore/Swift/Tests/API/BasicCompileTests.swift:507]: Done

File: Firestore/Swift/Source/AsyncAwait/Firestore+AsyncAwait.swift [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import FirebaseFirestore
+import Foundation
+
+#if compiler(>=5.5) && canImport(_Concurrency)
+  @available(iOS 15, tvOS 15, macOS 12, watchOS 8, *)
+  public extension Firestore {
+    /// Loads a Firestore bundle into the local cache.
+    /// - Parameter bundleData: Data from the bundle to be loaded.
+    /// - Returns: The final `LoadBundleTaskProgress` that contains the total number of documents loaded.
+    func loadBundle(_ bundleData: Data) async throws -> LoadBundleTaskProgress {
```
Comment (paulb777) [inline Firestore/Swift/Source/AsyncAwait/Firestore+AsyncAwait.swift:26]: Should there also be a `loadBundleStream`?

**Review by paulb777 — COMMENTED** [review]
All changes LGTM pending one more question.

Approval is also pending an API approval. cc: @ryanwilson

File: Firestore/Swift/Source/AsyncAwait/Firestore+AsyncAwait.swift (hunk 6 above)
Comment (schmidt-sebastian) [reply Firestore/Swift/Source/AsyncAwait/Firestore+AsyncAwait.swift:26]: Done as well.

File: Firestore/Swift/Source/AsyncAwait/Firestore+AsyncAwait.swift [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+import FirebaseFirestore
+import Foundation
+
+#if compiler(>=5.5) && canImport(_Concurrency)
+  @available(iOS 15, tvOS 15, macOS 12, watchOS 8, *)
+  public extension Firestore {
+    /// Loads a Firestore bundle into the local cache.
+    /// - Parameter bundleData: Data from the bundle to be loaded.
+    /// - Returns: The final `LoadBundleTaskProgress` that contains the total number of documents loaded.
```
Comment (ryanwilson) [inline Firestore/Swift/Source/AsyncAwait/Firestore+AsyncAwait.swift:25]: Nit: can you also describe what type of error is thrown, and the conditions that may cause it to fail? https://developer.apple.com/documentation/xcode/writing-symbol-documentation-in-your-source-files#Describe-the-Thrown-Errors-of-a-Method

File: Firestore/Swift/Source/AsyncAwait/Firestore+AsyncAwait.swift [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+          } else {
+            continuation.resume(returning: progress!)
+          }
+        }
+      }
+    }
+
+    /// Loads a Firestore bundle into the local cache.
+    /// - Parameter bundleStream: An input stream from which the bundle can be read.
+    /// - Returns: The final `LoadBundleTaskProgress` that contains the total number of documents loaded.
```
Comment (ryanwilson) [inline Firestore/Swift/Source/AsyncAwait/Firestore+AsyncAwait.swift:41]: Same here re: throwing type

File: Firestore/Swift/Tests/Integration/AsyncAwaitIntegrationTests.swift [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+      XCTAssertEqual(LoadBundleTaskState.success, bundleProgress.state)
+    }
+
+    func testLoadBundleFromStream() async throws {
+      let bundle = "\(emptyBundle.count)\(emptyBundle)"
+      let bundleProgress = try await db
+        .loadBundle(InputStream(data: bundle.data(using: String.Encoding.utf8)!))
+      XCTAssertEqual(LoadBundleTaskState.success, bundleProgress.state)
+    }
+  }
```
Comment (ryanwilson) [inline Firestore/Swift/Tests/Integration/AsyncAwaitIntegrationTests.swift:58]: 😍  my goodness these tests are easy to read... wonderful.

File: Firestore/Swift/Tests/Integration/AsyncAwaitIntegrationTests.swift [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+      let bundle = "\(emptyBundle.count)\(emptyBundle)"
+      let bundleProgress = try await db.loadBundle(Data(bundle.utf8))
+      XCTAssertEqual(LoadBundleTaskState.success, bundleProgress.state)
+    }
+
+    func testLoadBundleFromStream() async throws {
+      let bundle = "\(emptyBundle.count)\(emptyBundle)"
+      let bundleProgress = try await db
+        .loadBundle(InputStream(data: bundle.data(using: String.Encoding.utf8)!))
+      XCTAssertEqual(LoadBundleTaskState.success, bundleProgress.state)
```
Comment (ryanwilson) [inline Firestore/Swift/Tests/Integration/AsyncAwaitIntegrationTests.swift:56]: Can we also test the failure case, trying to load a a bundle that's empty `Data()`?

**Review by ryanwilson — COMMENTED** [review]
Overall LGTM!

File: Firestore/Swift/Source/AsyncAwait/CollectionReference+AsyncAwait.swift [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+#if compiler(>=5.5) && canImport(_Concurrency)
+  @available(iOS 15, tvOS 15, macOS 12, watchOS 8, *)
+  public extension CollectionReference {
+    /// Adds a new document to this collection with the specified data, assigning it a document ID
+    /// automatically.
+    /// - Parameter data: A `Dictionary` containing the data for the new document.
+    /// - Returns: A `DocumentReference` pointing to the newly created document.
+    func addDocument(data: [String: Any]) async throws -> DocumentReference {
+      typealias DataContinuation = CheckedContinuation<DocumentReference, Error>
```
[the remark this answers is not in the document]
Comment (paulb777) [reply Firestore/Swift/Source/AsyncAwait/CollectionReference+AsyncAwait.swift:28]: The `typealias` is the pattern I used in FirebaseStorageSwift/Sources/AsyncAwait.swift that I got from a reference that I can no longer find and it's not necessary. I made #9159 to fix Storage

File: Firestore/Swift/Source/AsyncAwait/CollectionReference+AsyncAwait.swift [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+      return try await withCheckedThrowingContinuation { (continuation: DataContinuation) in
+        var document: DocumentReference?
+        document = self.addDocument(data: data) { error in
+          if let err = error {
+            continuation.resume(throwing: err)
+          } else {
+            continuation.resume(returning: document!)
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply Firestore/Swift/Source/AsyncAwait/CollectionReference+AsyncAwait.swift:36]: Comment added

File: Firestore/Swift/Source/AsyncAwait/Firestore+AsyncAwait.swift (hunk 7 above)
Comment (schmidt-sebastian) [reply Firestore/Swift/Source/AsyncAwait/Firestore+AsyncAwait.swift:26]: Done - but I am not quite sure how to specify what type of error we throw since we just throw an `NSError` with a custom error domain.

File: Firestore/Swift/Source/AsyncAwait/Firestore+AsyncAwait.swift (hunk 8 above)
Comment (schmidt-sebastian) [reply Firestore/Swift/Source/AsyncAwait/Firestore+AsyncAwait.swift:43]: Done

File: Firestore/Swift/Tests/Integration/AsyncAwaitIntegrationTests.swift (hunk 10 above)
Comment (schmidt-sebastian) [reply Firestore/Swift/Tests/Integration/AsyncAwaitIntegrationTests.swift:66]: Done

**Review by schmidt-sebastian — COMMENTED** [review]
Feedback addressed. Sorry for the slow response time.

**Review by ryanwilson — APPROVED** [review]
Thanks!
